computed view endpoint for api client (GGT-7201)

mkobetic opened the pull request:
This is the counterpart to https://github.com/gadget-inc/gadget/pull/13298

It adds operation builder and runner for computed view as well as generic ComputeView function types and interfaces with metadata. It introduces two variants "with variables" and "without variables" to keep things simpler. The types probably could be merged, but I had enough trouble sorting the types out as is, we can try in a follow up.

Chunk of the work was debugging and tweaking the various utility types to allow TS to infer types correctly. It was a crazy learning curve for me, it's quite possible that I still got some things wrong here. I introduced second versions of the utility types as some of the tweaks broke the dependent packages that use the original ones (react). Not all tweaks are likely necessary for the purposes of this particular PR, but they all are addressing actual issues in the older types, so probably worth keeping. I'll explain more about each inline.

If this passes the review I'll need some guidance on how to roll this out.

## PR Checklist

- [ ] Important or complicated code is tested
- [ ] Any user facing changes are documented in the Gadget-side changelog
- [ ] Any immediate changes are slated for release in Gadget via a generated package dependency bump
- [ ] Versions within this monorepo are matching and there's a valid upgrade path

mkobetic commented on packages/api-client-core/src/GadgetFunctions.ts:
```diff
@@ -206,3 +207,14 @@ export interface GlobalActionFunction<VariablesT> {
 export type AnyActionFunction = ActionFunctionMetadata<any, any, any, any, any, any> | GlobalActionFunction<any>;
 export type AnyBulkActionFunction = ActionFunctionMetadata<any, any, any, any, any, true>;
+
+export interface ComputedViewFunction<VariablesT extends VariablesOptions, SelectionsT extends FieldSelection, ResultT> {
```
Let's use DefaultSelection, allows restricting it to fewer fields.

mkobetic commented on packages/api-client-core/src/GadgetFunctions.ts:
```diff
@@ -206,3 +207,14 @@ export interface GlobalActionFunction<VariablesT> {
 export type AnyBulkActionFunction = ActionFunctionMetadata<any, any, any, any, any, true>;
+
+export interface ComputedViewFunction<VariablesT extends VariablesOptions, SelectionsT extends FieldSelection, ResultT> {
+  (variables: VariablesT): Promise<ResultT>;
```
(variables: VariablesT, options?: { select?: SelectionsT }): Promise<Select<ResultT, DefaultSelection<DefaultSelectionsT, SelectionsT, DefaultSelectionsT>>>

mkobetic replied:
@airhorns so I gave this a go and it seems I have to declare this as follows for TS to stop complaining about things:
```ts
export interface ComputedViewFunction<VariablesT, DefaultSelectionsT extends FieldSelection, SelectionsT extends DefaultSelectionsT, ResultT> {
  (variables: VariablesT, options?: { select?: SelectionsT }): Promise<Select<ResultT, DefaultSelection<DefaultSelectionsT, { select?: SelectionsT | null }, DefaultSelectionsT>>>;
  /// ...
}
```
The `extends FieldSelection` bit is required because `Select<a,b>` requires b to extend FieldSelection. So far so good. The part that doesn't make sense to me is that I had to do `SelectionsT extends DefaultSelectionsT` because `DefaultSelection<a,b,c>` requires b and c to `extend a`. But that seems backwards, DefaultSelectionsT has the full selection details and SelectionT has some subset so it should be `DefaultSelectionT extends SelectionT` no? But if I do that I can't make DefaultSelection<> to work. This is all very meta and makes my head hurt, so quite likely I'm misunderstanding something.

mkobetic replied:
Hm, maybe I misunderstood what the DefaultSelectionT type should be, looking at this example https://github.com/gadget-inc/js-clients/blob/main/packages/api-client-core/spec/TestSchema.ts#L75. It seems that DefaultSelectionT should have all fields, but optional. So in that sense SelectionT could be extension of DefaultSelectionT. I thought DefaultSelectionT would have all fields required.

airhorns replied:
Thats correct -- the existing default selections have everything optional but I think thats really just happenstance more than anything, no strong reason they are actually that way. The *actual* selection type has to have all the keys optional so that users can pass only some keys and not have to exhaustively list `false` for everything else and I bet the default just matches that cause that was our headspace at the . Because they are out in the wild I think any new ones you produce should match methinks 👍

mkobetic replied:
I think with my improved understanding and leaning heavily on the TestSchema example I'm inclined to settle on this:
```ts
export interface ComputedViewFunction<
  Variables,
  AvailableSelection extends FieldSelection,
  Result
> {
  (variables: Variables, options?: { select?: AvailableSelection }): Promise<
    Select<Result, DefaultSelection<AvailableSelection, Selectable<AvailableSelection>, Required<AvailableSelection>>>
  >;
    // ...
}
```
I was struggling with the two Selection types, given that we can't really know what the actual selection is going to be at runtime. So I think it's fine to say that if the select: option is present the value is of type AvailableSelection and the default is a Required<> derivative of AvailableSelection. Does that make sense?

mkobetic replied:
Hrm, still something off, this give me `never`
```
Select<TotalInStock, DefaultSelection<AvailableTotalInStockSelection, Selectable<AvailableTotalInStockSelection>, Required<AvailableTotalInStockSelection>>>
```

mkobetic replied:
I'm thinking the issue is with the default arg `Required<AvailableSelection>`, we don't want just required fields, we want them all set to true, so I need a type that has all fields and all typed as 'true' I think.

mkobetic commented on packages/api-client-core/spec/default-selection.spec.ts:
```diff
@@ -1,11 +1,90 @@
 
-type _NonNullDefault = DefaultSelection<AvailableTestSchemaSelection, { select: { num: false; str: true } }, { num: true }>;
-type _TestRespectsTruthySelections = AssertTrue<IsExact<_NonNullDefault, { num: false; str: true }>>;
+  type undefinedDefault = DefaultSelection<AvailableTestSchemaSelection, { select: undefined }, { num: true }>;
```
adding more cases where the default should result.

mkobetic commented on packages/api-client-core/spec/default-selection.spec.ts:
```diff
@@ -1,11 +1,90 @@
+  type undefinedDefault2 = DefaultSelection<AvailableTestSchemaSelection, {}, { num: true }>;
+  type _TestDefaultsUndefinedToTheDefault2 = AssertTrue<IsExact<undefinedDefault2, { num: true }>>;
+
+  type anyDefault = DefaultSelection2<AvailableTestSchemaSelection, { select: any }, { num: true }>;
```
Again `any` breaks the original DefaultSelection.

mkobetic commented on packages/api-client-core/spec/default-selection.spec.ts:
```diff
@@ -1,11 +1,90 @@
+  type _TestAvailableTestSchemaSelection = Has<FieldSelection, availableTestSchemaSelection>;
+
+  type testType = { a: number; b: { c: string; d: { e: boolean } } };
+  type availableSelection = AvailableSelection2<testType>;
```
The original AvailableSelection has a bug where boolean fields seem to end up being just `boolean | undefined` instead of `boolean | null | undefined`.

mkobetic commented on packages/api-client-core/src/types.ts:
```diff
@@ -21,15 +21,66 @@ export type LimitToKnownKeys<T, U> = {
+  ? boolean | null | undefined
+  : { [key in keyof Schema]?: AvailableSelection<Schema[key]> };
+
+export type AvailableSelection2<Schema> = Schema extends Array<infer T>
```
Two issues with the original:
* doesn't handle arrays properly
* converts boolean fields into `boolean | undefined` instead of `boolean | null | undefined` for some reason

airhorns replied:
Why introduce a copy instead of just changing the original? Some backwards compatibility concern?

airhorns replied:
I see I see. To me that suggests that either the react package needs to change and never pass in `any`, or that this change is actually a bad idea and that `any` occurs "naturally" as an input type for good reasons. Can you hunt down if we want to eliminate `any` entirely or not

mkobetic replied:
I can definitely dig into that, I'm just not sure how much of a rabbit hole that's going to be. Does it make sense to ship this with the duplication and then follow up with the attempted merge? Otherwise I'm not sure how much more this is going to drag out.

airhorns replied:
We can merge without the deduplication sure! I think its worth resolving before release though.

mkobetic commented on packages/api-client-core/src/types.ts:
```diff
@@ -21,15 +21,66 @@ export type LimitToKnownKeys<T, U> = {
   Defaults extends SelectionType
 > = Options["select"] extends SelectionType ? Options["select"] : Defaults;
 
+export type DefaultSelection2<
```
This one I think I do need for the original PR, tightening the types likely allows the TS inference to do the right thing for me (not 100% sure but at this point I just want to get the whole thing into a working state before tweaking things further). It also fixes it to return the Defaults when the Options ends up resolved to `any` instead of returning `any` which screws everything up.

This is also likely the one that breaks the other packages like `react` because the types are now a lot tighter than before. Again would need to dig more to confirm.

mkobetic commented on packages/api-client-core/src/types.ts:
```diff
@@ -74,6 +140,24 @@ type InnerSelect<Schema, Selection extends FieldSelection | null | undefined> =
         : never;
     };
 
+type InnerSelect2<Schema, Selection extends FieldSelection | null | undefined> = IfAny<
```
The fix here is to return `never` when Selection ends up being `any`. I probably don't need this for the PR, but we should probably fix this regardless.

mkobetic commented on packages/api-client-core/src/types.ts:
```diff
@@ -749,33 +834,6 @@ export type PaginateOptions = {
- */
-export type AvailableSelection<Schema> = Schema extends string | number | bigint | null | undefined
-  ? boolean | null | undefined
-  : { [key in keyof Schema]?: AvailableSelection<Schema[key]> };
```
I just moved this up to be among the generic utility types.

mkobetic commented on packages/api-client-core/src/GadgetFunctions.ts:
```diff
@@ -206,3 +207,63 @@ export interface GlobalActionFunction<VariablesT> {
+  // Available, Options and Defaults are inferred at call time.
+  <
+    // Available is the full FieldSelection type derived from the shape of the Result type, i.e. all possible selections.
+    Available extends AvailableSelection2<Result> & FieldSelection,
```
Note the `& FieldSelection`, something about `AvailableSelection<>` makes it not assignable to `FieldSelection` so narrowing it down to what matches.

airhorns commented on packages/api-client-core/spec/operationBuilders.spec.ts:
```diff
@@ -1491,4 +1492,52 @@ describe("operation builders", () => {
+        {
+          "query": "query boom($a: Int, $b: String) {
+          boom(a: $a, b: $b) {
+            a
```
👍

airhorns reviewed (approved):
This LGTM after the question about the copied types!

mkobetic commented:
I added some tests and squashed and rebased and apparently I can't build the package anymore for some reason. The error doesn't seem to have anything to do with what I've changed.

```
martin@MacBookPro api-client-core % npm run build

> @gadgetinc/api-client-core@0.15.33 build
> rm -rf dist && mkdir -p dist/cjs dist/esm && echo '{"type": "commonjs"}' > dist/cjs/package.json && echo '{"type": "module"}' > dist/esm/package.json && tsc -b tsconfig.cjs.json tsconfig.esm.json && tsx build/replace-version-constant.ts

src/GadgetConnection.ts:500:47 - error TS2554: Expected 3 arguments, but got 2.

500           this.subscriptionClientOptions?.on?.connected?.(socket, payload);
                                                  ~~~~~~~~~

  node_modules/graphql-ws/lib/client.d.ts:84:98
    84 export type EventConnectedListener = (socket: unknown, payload: ConnectionAckMessage['payload'], wasRetry: boolean) => void;
                                                                                                        ~~~~~~~~~~~~~~~~~
    An argument for 'wasRetry' was not provided.

src/GadgetConnection.ts:501:26 - error TS2554: Expected 3 arguments, but got 2.

501           overrides?.on?.connected?.(socket, payload);
                             ~~~~~~~~~

  node_modules/graphql-ws/lib/client.d.ts:84:98
    84 export type EventConnectedListener = (socket: unknown, payload: ConnectionAckMessage['payload'], wasRetry: boolean) => void;
                                                                                                        ~~~~~~~~~~~~~~~~~
    An argument for 'wasRetry' was not provided.

src/GadgetConnection.ts:500:47 - error TS2554: Expected 3 arguments, but got 2.

500           this.subscriptionClientOptions?.on?.connected?.(socket, payload);
                                                  ~~~~~~~~~

  node_modules/graphql-ws/lib/client.d.ts:84:98
    84 export type EventConnectedListener = (socket: unknown, payload: ConnectionAckMessage['payload'], wasRetry: boolean) => void;
                                                                                                        ~~~~~~~~~~~~~~~~~
    An argument for 'wasRetry' was not provided.

src/GadgetConnection.ts:501:26 - error TS2554: Expected 3 arguments, but got 2.

501           overrides?.on?.connected?.(socket, payload);
                             ~~~~~~~~~

  node_modules/graphql-ws/lib/client.d.ts:84:98
    84 export type EventConnectedListener = (socket: unknown, payload: ConnectionAckMessage['payload'], wasRetry: boolean) => void;
                                                                                                        ~~~~~~~~~~~~~~~~~
    An argument for 'wasRetry' was not provided.


Found 4 errors.

```

airhorns commented:
Hm, the fact that it is green in CI makes me think its something strange about your local environment! Can I suggest `rm -rf node_modules; pnpm install` to see if its some corruption in there?

mkobetic commented:
> Hm, the fact that it is green in CI makes me think its something strange about your local environment! Can I suggest `rm -rf node_modules; pnpm install` to see if its some corruption in there?

Yup, looks like wiping the modules helped. I also noticed I managed to mix pnpm and npm, maybe that's what screwed things up.
